Refresh token

TEdwardJ opened the pull request.

TEdwardJ committed:
Refresh Token:
 - UserSecurityResponse is extended in order to return the client refresh token along with access token;

TEdwardJ committed:
Refresh Token:
 - new endpoint /refresh that lets the client obtain new access and refresh tokens pair;
 - response of /login includes refreshToken;
 - JwtService.createRefreshToken() method declared;

TEdwardJ committed:
Merge remote-tracking branch 'origin/release_25082021' into refresh_token

TEdwardJ committed:
Refresh Token Functionality:
 - new endpoint /token/refresh-token added into JwtLoginController;
 - since the UserSecurityResponse class is used to prepare response for user`s login request , it has got new field refreshToken;
 - new class named TokenPair is added in order to keep refresh and access logically grouped with the same pair id;
 - LogoutService and InMemoryLogoutTokenService has got new functionality in order to track used tokenPairs;
 - JwtTokenProvider rework;
 - tests that corresponding to the specified changes are reworked;

TEdwardJ committed:
Refresh Token Functionality:
 - check token endpoind (/token) should return 401 HttpStatus in case of expired token;

TEdwardJ committed:
Refresh Token Functionality:
 - temporarily decreasing of jwt.token.expirationInSeconds to 20 seconds;

TEdwardJ committed:
Refresh Token Functionality:
 - 401 http status for token expiration case;

TEdwardJ committed:
UniqueVinyl:
 - All Args Constructor is added, boolean field has got Boolean type;

TEdwardJ committed:
UniqueVinyl:
 - offers boolean field renamed to hasOffers due to ElasticSearch existing database limitations;

TEdwardJ committed:
Refresh Token Functionality:
 - ApiExceptionHandler - added handler for AccessDeniedException;
 - JwtTokenProvider.refresh() - refresh logic should be performed only in case the token is valid and the user the token is related to exists;
  - HttpStatus 401 instead of 403 in case of expired jwt Token;

TEdwardJ committed:
Refresh Token Functionality:
 - Tests for profile controller are fixed;

TEdwardJ committed:
Refresh Token Functionality:
 - unused imports elimination;
 - obsolete variables elimination;

littlepony99 commented on src/main/java/com/vinylteam/vinyl/entity/TokenPair.java:
```diff
@@ -0,0 +1,13 @@
+
+    String id;
+    String jwtToken;
+    String refreshToken;
```
add empty row after last class row

littlepony99 commented on src/main/java/com/vinylteam/vinyl/security/LogoutTokenStorageService.java:
```diff
@@ -3,7 +3,8 @@
-    boolean isTokenBlocked(String token);
+    boolean isTokenPairBlocked(String pairIdentifier);
+
+    void storePairIdentifier(String pairIdentifier, LocalDateTime expirationDate);
```
add empty row after last class row

littlepony99 commented on src/main/java/com/vinylteam/vinyl/security/SecurityConfiguration.java:
```diff
@@ -142,7 +142,7 @@ public LogoutService logoutService(JwtService jwtService,
         DelegatingFilterProxy delegateFilterProxy = new DelegatingFilterProxy();
         delegateFilterProxy.setServletContext(context);
-        delegateFilterProxy.setTargetBeanName("jwtValidationFilter");
+        delegateFilterProxy.setTargetBeanName("jwtValidatorFilter");
```
move to a constant variable?

littlepony99 commented on src/main/java/com/vinylteam/vinyl/service/DefaultProfileManagementService.java:
```diff
@@ -43,7 +45,7 @@ public ChangePasswordResponse changeProfilePassword(HttpServletRequest request,
                 .authenticateByRequest(new LoginRequest(user.getEmail(), userProfileInfo.getNewPassword()))
-                .getToken();
+                .getJwtToken();
         logoutService.logout(request, null, null);
```
seems to me that its not a good "interface" idea
maybe you should create logout method with 1 param, which under hood will send request logoutService.logout(request, null, null);

littlepony99 commented on src/main/java/com/vinylteam/vinyl/service/JwtService.java:
```diff
@@ -1,28 +1,41 @@
+    TokenPair getTokenPair(JwtUser user);
+
+    UserSecurityResponse refreshByToken(String refreshToken);
 }
```
add empty row after last class row

littlepony99 commented on src/main/java/com/vinylteam/vinyl/service/impl/DefaultUserService.java:
```diff
@@ -26,9 +27,10 @@
 
+    private SecurityService securityService;
+
+    private final EmailConfirmationService emailConfirmationService;
```
add empty row, divide variable and method parts

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/entity/Currency.java:
```diff
@@ -5,6 +5,8 @@
 
 import java.util.Optional;
```
please remove all logging statements from this enum

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/entity/Currency.java:
```diff
@@ -5,6 +5,8 @@
 
 import java.util.Optional;
```
could you please also check this code?

else if (" EUR".equals(currencyDescription) || "EUR".equals(currencyDescription) || "€".equals(currencyDescription)
                || "&nbsp;€".equals(currencyDescription)) {
            resultingCurrency = EUR;
        }

is white space between " and EUR ok ?

" EUR .."

TEdwardJ replied:
Yes, I suppose the author wanted to be sure that any possible string with currency will be recognized correctly, but think trim() will be better

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/entity/TokenPair.java:
```diff
@@ -0,0 +1,13 @@
+@Builder
+public class TokenPair {
+
+    String id;
```
why package-private modifiers?

TEdwardJ replied:
@FieldDefaults(makeFinal=true, level=AccessLevel.PRIVATE)  - defined in the @Value annotation

TEdwardJ replied:
@FieldDefaults(makeFinal=true, level=AccessLevel.PRIVATE) from @Value annotation from lombok

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/entity/UniqueVinyl.java:
```diff
@@ -27,16 +28,11 @@
     @Field(type = FieldType.Boolean)
     @EqualsAndHashCode.Exclude
-    private boolean offers;
+    private boolean hasOffers;
```
please revert this
its should be "offers"
we use has* for boolean's getters in this case

TEdwardJ replied:
it requires reloading data to ElasticSearch... or probably there is another solution, but it still has to be found. Just renaming makes the catalog page fails

hemiSRT8 commented on src/main/java/com/vinylteam/vinyl/security/InMemoryLogoutTokenService.java:
```diff
@@ -27,7 +27,7 @@ public void initCache() {
     }
 
     private Expiry<String, LocalDateTime> getExpirationStrategy() {
```
please move this private method below public ones

hemiSRT8 replied:
could you please also fix all parameters String s in this class?
you could go to the documentation of com.github.benmanes.caffeine.cache.Expiry and read what s means

TEdwardJ committed:
Refresh Token PR review:
 - empty lines added to follow the code style;
 - LogoutService renamed to DefaultLogoutService;
  - new interface added, named LogoutService;

TEdwardJ committed:
Refresh token PR review:
 - private method to the bottom of the class declaration;
 - Currency class has the currency recognizing logic changed;
